### PR TITLE
sql/stats: support rowCountEq = 0 in histogram.adjustCounts

### PR DESCRIPTION
**Pull request by michae2**
The predicted histograms in statistics forecasts will often have buckets
with NumEq = 0, and some predicted histograms will have _all_ buckets
with NumEq = 0. This wasn't possible before forecasting, because the
histograms produced by `EquiDepthHistogram` never have any buckets with
NumEq = 0.

If `adjustCounts` is called on such a histogram, `rowCountEq` and
`distinctCountEq` will be zero. `adjustCounts` should still be able to
fix such a histogram to have sum(NumRange) = rowCountTotal and
sum(DistinctRange) = distinctCountTotal. This patch teaches
`adjustCounts` to handle these histograms.

(Similarly, predicted histograms could have all buckets with
NumRange = 0, but this is already possible for histograms produced by
`EquiDepthHistogram`, so `adjustCounts` already handles these.)

Also, add a few more comments to `adjustCounts`.

Assists: #79872

Release note: None

### CODE REVIEW

**Review by rytaft — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82474#-:-N3uCAuH0qKitk08w2fA:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @msirek)

File: pkg/sql/stats/histogram.go [hunk 1]
```diff
@@ -170,8 +170,16 @@ type histogram struct {
 }
 
 // adjustCounts adjusts the row count and number of distinct values per bucket
-// based on the total row count and estimated distinct count.
+// to equal the total row count and estimated distinct count. The total row
+// count and estimated distinct count should not include NULL values, and the
+// histogram should not contain any buckets for NULL values.
 func (h *histogram) adjustCounts(evalCtx *eval.Context, rowCountTotal, distinctCountTotal float64) {
+	// Empty table cases.
+	if rowCountTotal <= 0 || distinctCountTotal <= 0 {
```
Comment (mgartner) [inline pkg/sql/stats/histogram.go:178]: I'm guessing negative counts are possible with forecasting too? Maybe the commit message should mention that. And you could add test cases with negative counts if you'd like.

**Review by mgartner — APPROVED** [review]
LGTM

**Review by msirek — APPROVED** [review]
:lgtm: 

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82474#-:-N3vGTMQCjvzl8nnzKIG:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @michae2)

File: pkg/sql/stats/histogram.go (hunk 1 above)
Comment (michae2) [reply pkg/sql/stats/histogram.go:178]: Good point, I added a test case.

The current PR [normalizes negative counts to zero](https://github.com/cockroachdb/cockroach/blob/c903654bbbb2d351625e28a5864766b5a83f8af6/pkg/sql/stats/forecast.go#L261) before calling `adjustCounts` but you're absolutely right, this was my intention with <=.

**Comment by michae2** [comment]
TFTRs!

bors r=rytaft,mgartner,msirek
